gossipd/gossip_store.c

```diff
@@ -96,58 +97,69 @@ void gossip_store_add_channel_delete(struct gossip_store *gs,
 	tal_free(msg);
 }
 
-bool gossip_store_read_next(struct routing_state *rstate,
-			    struct gossip_store *gs)
+void gossip_store_load(struct routing_state *rstate, struct gossip_store *gs)
 {
 	beint32_t belen;
 	u32 msglen;
 	u8 *msg, *gossip_msg;
 	u64 satoshis;
-	enum gossip_wire_type type;
 	struct short_channel_id scid;
-
-	/* Did we already reach the end of the gossip_store? */
-	if (gs->read_pos == -1)
-		return NULL;
-
-	/* Can we read one message? */
-	if (pread(gs->fd, &belen, sizeof(belen), gs->read_pos) != sizeof(belen)) {
-		gs->read_pos = -1;
-		return false;
-	}
-
-	msglen = be32_to_cpu(belen);
-	msg = tal_arr(gs, u8, msglen);
-
-	if (!pread(gs->fd, msg, msglen, gs->read_pos + sizeof(belen))) {
-		status_trace("Short read from gossip-store, expected lenght %d",
-			     msglen);
-
-		/* Reset write_pos to truncate this message and disable future
-		 * reads */
-		gs->write_pos = gs->read_pos;
-		gs->read_pos = -1;
-		ftruncate(gs->fd, gs->write_pos);
+	/* We set/check version byte on creation */
+	off_t known_good = 1;
+	const char *bad;
+
+	lseek(gs->fd, known_good, SEEK_SET);
+	while (read(gs->fd, &belen, sizeof(belen)) == sizeof(belen)) {
+		msglen = be32_to_cpu(belen);
+		msg = tal_arr(gs, u8, msglen);
+
+		if (read(gs->fd, msg, msglen) != msglen) {
+			status_unusual("gossip_store: truncated file?");
+			goto truncate_nomsg;
+		}
+
+		if (fromwire_gossip_store_channel_announcement(msg, msg,
```

Comment:
Any reason we are moving away from the `peektype` / `switch` kind of routing?

Reply:
It's done internally in those functions, so it was redundant.